### DISCUSSION TITLE
Use Weblate for translations

### DISCUSSION THREAD

**Discussion by ignotus666** [jamulussoftware/jamulus#2675] [Ideas]
@jamulussoftware/translators @ann0see @gilgongo 

Though the po4a setup made things a bit simpler, I think there are still big barriers when it comes to translating the documentation. A big one is having to navigate all the GitHub stuff (forking, branches, PRs, conflicts, etc.). Further problems arise with local translating due to different OS's, editors/versions, incorrect settings, etc. A case in point is the ongoing German translation, where this poor person had to jump through countless hoops and waste a lot of time.

I had another look at [Weblate](https://weblate.org/en-gb/). Back in the day when I checked it out I found it quite daunting and also failed to see that there is an option of free hosting for libre projects (you otherwise need to either pay or host the service yourself). I've set up a trial account linked to my repo and have been doing some testing, and it looks quite promising - basically what it does is shift the translation process itself to an online interface. Translators only need a GitHub account and don't need to worry about forks, PRs or any of that, as Weblate itself creates forks and you can configure it to automatically push commits, create PRs, or push manually. It also pulls changes to keep in sync with the repo. There are nice features such as highlighting source file changes, checking for spacing, line break and other errors, displaying translation progress charts on your repo, and a lot of other stuff I still need to get my head around. Incidentally, the po4a project itself uses Weblate.
Translators just have to log in to Weblate, click through to the documents they want to translate and do their translations there, the rest is taken care of, either automatically, or by admins. The po4a setup and all the GH actions on our repo would stay as they are.

I reckon it'd be a bit premature to roll it out for this release and first I wanted to see what you think. The project needs to qualify for free hosting, but I think we meet the requirements (you can see them [here](https://hosted.weblate.org/hosting/)).

You can check out the trial version I set up [here](https://hosted.weblate.org/projects/ignotus/).

**Comment by ignotus666** [comment]
I just realised this might be better in Discussions... feel free to move it.

**Reply by ann0see** [reply]
Done since it might also apply to the app

**Comment by ann0see** [comment]
Hmm yes after looking at the page it looks quite promising. I'd however prefer not needing to have another platform to maintain

**Comment by ignotus666** [comment]
True, but conversely it could save a lot of hand-holding and make the translation process much more accessible by removing the requirement of needing to know how to use GitHub.

**Comment by BLumia** [comment]
> True, but conversely it could save a lot of hand-holding and make the translation process much more accessible by removing the requirement of needing to know how to use GitHub.

Additionally, Weblate also send translation updates via GitHub Pull Request automatically, so if choosing Weblate, we can still accept regular GitHub Pull Request if a translator doesn't want to use the hosted Weblate instance.

One thing might be worth to mention is, if we choose to use Weblate, we'll need to pay more attention to the source strings that contains arguments (e.g. `"About %1"`) and make use of the 2nd `disambiguation` argument in `QObject::tr()` to tell translator what's the argument means, since unlike using QtLinguist with full source code fetched, the translation platform won't show the related source code in the translation page.

I'm using Weblate for my tiny side project and so far it's pretty good for me. You can check out [here](https://hosted.weblate.org/projects/pineapple-pictures/application/) for what it looks like for the Weblate translation process (how to add a new language, how to translate a string, etc.), and [here](https://github.com/BLumia/pineapple-pictures/pull/54) for a Pull Request that automatically sent by Weblate to update the ts files.

HTH

**Comment by ann0see** [comment]
Yeah. Valid point. I'd say it's up to @jujudusud @henkdegroot too if they want to use it. I used transifex some time ago (or however it's called) and it was convenient.

**Reply by jujudusud** [reply]
Ok for me to use another tool, interface or ... ;-)

**Comment by henkdegroot** [comment]
The current setup works fine for me. Not against using anything else, however when I do translation(s) I usually want to be able to "see" the end result before submitting a PR. Hence I already need to install/download code etc. 

Having a less github involved setup could benefit to attract new translators.

**Comment by trebmuh** [comment]
Hi all. I use Weblate for a bunch of other projects I translate and I :+1: this proposal. Weblate is very convenient for translators because it removes all the "non-translation" barriers (code,...).
One can use github to connect to it or use a login/password if he doesn't have a github account.
Using Weblate is both a good and a bad thing for having more people participating to the translation process and I recommend you to set up one or two "reviewer" for languages with more than one translator. Otherwise, you might have editing conflicts between people.

By the way, this would probably allow me to participate in the jamulus translation again.

**Comment by ann0see** [comment]
Ok. So I'm not going to stop you from implementing it. However no guarantee for being online there.

If it's beneficial for both the website and the app, setting it up for both would also be good.

**Reply by ignotus666** [reply]
I just added the app .ts file to my trial weblate account. If anyone wants to play around with it just let me know and I'll add you as a contributor.

**Reply by trebmuh** [reply]
@ignotus666 please, add me (hosted.weblate user "trebmuh")

**Reply by ignotus666** [reply]
@trebmuh  Done. I've added you as a translator, but I can give you administrator rights if you want.

**Reply by trebmuh** [reply]
I translated 3 strings to check if that was working as usual. It looks to.
@jujudusud: you've been doing pretty much all the translations since I stopped after doing the initial work. I would suggest you and me being the 2 persons able to validate each translation. What do you reckon on this?

**Reply by jujudusud** [reply]
@trebmuh , it's ok for me.

**Comment by ignotus666** [comment]
@trebmuh do you manage any weblate accounts? There are things I think I need to understand better and it would be good to have someone with experience. Or @BLumia . OTOH, bear in mind this is just a test account, if we finally set one up for Jamulus it will be a new one.

**Reply by trebmuh** [reply]
@ignotus666 : no, I don't manage weblate accounts. I'm only using it as a translator (and sometime a "reviewer" like for zrythm as an example). You probably can ask alextee (he isn't on github AFAIK) who is the zrythm main dev and he had set up a hosted.weblate account for zrythm. There is an #zrythm IRC channel on libera.

Thanks for the "test" account precision, I suspected it. I'm not going to translate anything more on it for now and will wait an official Jamulus one if the move will happen.

**Reply by BLumia** [reply]
I also don't manage weblate accounts. I'm using it as a 3rd-party FOSS developer, and the things I can see/manage in my project should be basically the same as a trial project you've created.

**Reply by ignotus666** [reply]
@BLumia: Maybe I used the wrong terminology, I meant to ask whether you'd set up and manage an account - I take it you have for your Pineapple-Pictures project.
There are a few things that I see as potential annoyances, though hopefully someone with experience can shed some light on them:
- It seems you can't have separate PRs for each language. Commits to all languages on Weblate get bundled into a single PR here. In theory it shouldn't be a problem as the review would happen on Weblate, but maybe there are other downsides.
- ~~@trebmuh If we enforce reviews, does the reviewer really have to approve every single string one by one? I might have missed how you can "bulk approve" strings, but if you can't, that's pretty annoying.~~
- ~~These "Merge remote-tracking branch" commits seem to keep getting added without touching either the Jamulus repo or Weblate. But it could be a wrong setting somewhere related to how the Weblate account syncs to the repo.~~

*Edit: So there is actually a "bulk edit" tool to change the review status and the "Merge remote-tracking branch" commits are Weblate syncing after merging PRs in the repo. 
*Edit 2: PRs are separated by files ("components" in Weblate). So e.g. translations of "FAQ" are grouped into one PR, another PR for "Software-Manual", etc. So the app translations, for instance, would all go into a single, big PR.

OTOH, useful functions include:
- Glossary for consistent terminology.
- A load of pre-configured checks and the possibility of creating custom ones.
- Translation memory management.
- Can leave comments, suggestions, etc. on a string-by-string or project-wide basis.
- Can view history of all changes file-by-file or project-wide.
- Easy to revert changes.
- Lots of documentation.
- Easy to define roles for translators/reviewers/admins and what they can edit/access.
- Probably a lot more I haven't discovered yet...

**Comment by ann0see** [comment]
@ignotus666 I don't think we'll get this done for 3.9.0? I mean it would be probably good to have some testing by translators.

**Reply by ignotus666** [reply]
No, my idea for now was to see what people think and if it gets support we can use it for the following release. If people want to try it out my test installation is there but it'll expire in about 10 days - but anyway, the translation process itself is very straightforward, I think it's simpler than the current system. Where I see potential reticence is on the admin side, in how PRs are handled and the changes that BLumia mentions above.
But I understand there's a release process going on so I didn't want to insist on it yet.

**Comment by ann0see** [comment]
@ignotus666 Did you test Weblate some more? I think it's worth opening an issue and deciding on next steps since mostly responses were positive.

**Comment by ignotus666** [comment]
I did, up until it expired, and I still think that on the translators' side it would make thinks easier, particularly website translations. Also, if website and app translations were presented in the same place to translators, some might be tempted to contribute more to the website ;)

However, it would involve a few changes on the admin side (pointed out above) and I think that's what needs discussing. Shall I open a discussion then, outlining what would it would involve?

**Reply by ann0see** [reply]
Hmm. Yes.

**Comment by trebmuh** [comment]
@ignotus666 : can you please add me as a translator on weblate? I'm unable to make contribution so far. I'm only able to make some translation "suggestions".

**Reply by ignotus666** [reply]
@trebmuh it's locked for now, no one can translate until/unless we get approval for hosting.

**Reply by trebmuh** [reply]
@ignotus666 ah ok. I was under the impression that you had to give me permissions to do so. I misunderstood then.
By any chance, do you have any idea how long we have to wait before approval?

**Reply by ignotus666** [reply]
Well in theory I should get a reply before Monday, otherwise the Jamulus project I set up there will be deleted as the trial period ends. I've sent another message today but still no reply... Fingers crossed.

**Reply by trebmuh** [reply]
corsisng fnigres tehn
_(but not easy to type down on my keyboard then!)_

**Comment by ignotus666** [comment]
@trebmuh we've received approval and the project is unlocked, so you can now add your contributions.

**Reply by trebmuh** [reply]
Tried and everything seems to be working. Thanks you.

**Comment by ann0see** [comment]
Hmm. Just did a change on Weblate, but IDK if anyhing changed.

**Reply by jujudusud** [reply]
It's ok with your changes, right?

**Reply by ann0see** [reply]
Yes. But it still needs some investigation on my side.

**Reply by BLumia** [reply]
You might need to trigger a push from `Manager > Repository maintenance` then `Commit`

**Comment by gilgongo** [comment]
Hi @comradekingu - we'd like to invite you to our Discord. If you'd like to join, please email team@jamulus.io and ask for access. Thanks!
